server/src/main/java/io/unitycatalog/server/persist/ExternalLocationRepository.java

```diff
@@ -0,0 +1,195 @@
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+
+public class ExternalLocationRepository {
+  private final Repositories repositories;
+  private final SessionFactory sessionFactory;
+  private static final PagedListingHelper<ExternalLocationDAO> LISTING_HELPER =
+      new PagedListingHelper<>(ExternalLocationDAO.class);
+
+  public ExternalLocationRepository(Repositories repositories, SessionFactory sessionFactory) {
+    this.repositories = repositories;
+    this.sessionFactory = sessionFactory;
+  }
+
+  public ExternalLocationDAO addExternalLocation(CreateExternalLocation createExternalLocation) {
+    ValidationUtils.validateSqlObjectName(createExternalLocation.getName());
+    String callerId = IdentityUtils.findPrincipalEmailAddress();
+
+    try (Session session = sessionFactory.openSession()) {
+      StorageCredentialDAO storageCredentialDAO =
+          repositories
+              .getStorageCredentialRepository()
+              .getStorageCredentialDAO(session, createExternalLocation.getCredentialName());
+      UUID externalLocationId = UUID.randomUUID();
+      ExternalLocationDAO externalLocationDAO =
+          ExternalLocationDAO.builder()
+              .id(externalLocationId)
+              .name(createExternalLocation.getName())
+              .url(createExternalLocation.getUrl())
+              .comment(createExternalLocation.getComment())
+              .owner(callerId)
+              .accessPoint(createExternalLocation.getAccessPoint())
+              .credentialId(storageCredentialDAO.getId())
+              .createdAt(new Date())
+              .createdBy(callerId)
+              .build();
+      Transaction tx = session.beginTransaction();
+      try {
+        if (getExternalLocationDAO(session, createExternalLocation.getName()) != null) {
+          throw new BaseException(
+              ErrorCode.ALREADY_EXISTS,
+              "External location already exists: " + createExternalLocation.getName());
+        }
+        session.persist(externalLocationDAO);
+        tx.commit();
+        return externalLocationDAO;
+      } catch (Exception e) {
+        tx.rollback();
+        throw e;
+      }
+    }
+  }
+
+  public ExternalLocationInfo getExternalLocation(String name) {
+    try (Session session = sessionFactory.openSession()) {
+      session.setDefaultReadOnly(true);
+      Transaction tx = session.beginTransaction();
+      try {
+        ExternalLocationDAO externalLocationDAO = getExternalLocationDAO(session, name);
+        if (externalLocationDAO == null) {
+          throw new BaseException(ErrorCode.NOT_FOUND, "External location not found: " + name);
+        }
+        tx.commit();
+        return externalLocationDAO.toExternalLocationInfo();
+      } catch (Exception e) {
+        tx.rollback();
+        throw e;
+      }
+    }
+  }
+
+  public ExternalLocationDAO getExternalLocationDAO(Session session, String name) {
```

Comment:
why does this need to be public? why does the repo object need to expose the internal DAO objects? DAOs are basically internal details .. right?